### PR TITLE
Fixes to openssl compatibility for Boost.Asio with SSF

### PR DESCRIPTION
**Pull request by dgarske**
* Fixes for asio build options (so includes OPENSSL_EXTRA).
* Fix for bad named variable `shutdown`.
* Fix for the side size in Options struct to support `WOLFSSL_SIDE_NEITHER` (3).
* Fix to set the side on wolfSSL_connect() or wolfSSL_accept().
* Fix for `BIO_wpending` to work correctly.
* Fix for DH max size calc not including DH_Pub.
* Fix for PemToDer which was not properly handling extra new lines at end of file.

### CODE REVIEW

File: src/bio.c [hunk 1]
```diff
@@ -688,6 +687,31 @@ WOLFSSL_BIO* wolfSSL_BIO_next(WOLFSSL_BIO* bio)
...
+/* BIO_wpending returns the number of bytes pending to be written. */
+size_t wolfSSL_BIO_wpending(const WOLFSSL_BIO *bio)
+{
+    WOLFSSL_ENTER("BIO_wpending");
+
+    if (bio == NULL)
+        return 0;
+
+    if (bio->ssl != NULL) {
+        return (long)wolfSSL_pending(bio->ssl);
```
Comment (toddouska) [inline src/bio.c]: wolfSSL_pending() returns the number of buffered readable plaintext bytes.  Not writable.

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -7962,19 +7968,8 @@ int PemToDer(const unsigned char* buff, long longSz, int type,
...
-        return BUFFER_E;
-
-    /* eat end of line */
-    if (headerEnd[0] == '\n')
-        headerEnd++;
-    else if (headerEnd[1] == '\n')
-        headerEnd += 2;
-    else {
-        if (info)
-            info->consumed = (long)(headerEnd+2 - (char*)buff);
```
Comment (toddouska) [inline wolfcrypt/src/asn.c]: info->consumed marker was added for a reason, why are we removing it?

Comment (dgarske) [reply wolfcrypt/src/asn.c]: Its not removed, it get updated correctly in the lines below.

File: src/bio.c (hunk 1 above)
Comment (dgarske) [reply src/bio.c]: Okay I'll take a look at this again. Thanks

**Comment by dgarske** [comment]
@toddouska : Fixed the `wolfSSL_BIO_wpending` ssl case. Marked as not supported (also not used for the Boost.Asio project).

**Pull request merged**